test/e2e/run_test.sh

```diff
@@ -0,0 +1,128 @@
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+containerdConfigPatches:
+- |-
+  [plugins."io.containerd.grpc.v1.cri".registry]
+    config_path = "/etc/containerd/certs.d"
+EOF
+
+    status=$?
+    if [ $status -ne 0 ]; then
+        echo "Could not setup KinD environment. Something wrong with KinD setup. Exporting logs."
+    fi
+}
+
+export KIND_REGISTRY_NAME="kind-registry"
+export KIND_REGISTRY_PORT="5000"
+export KIND_REGISTRY="localhost:${KIND_REGISTRY_PORT}"
+
+# Provision a local docker registry, so KinD nodes could pull images from.
+# https://kind.sigs.k8s.io/docs/user/local-registry/
+function setup_kind_registry() {
+    running="$(docker inspect -f '{{.State.Running}}' "${KIND_REGISTRY_NAME}" 2>/dev/null || true)"
+    if [[ "${running}" != 'true' ]]; then
+        docker run \
+            -d --restart=always -p "${KIND_REGISTRY_PORT}:5000" --name "${KIND_REGISTRY_NAME}" \
+            gcr.io/istio-testing/registry:2
+        
+        # Allow kind nodes to reach the registry
+        docker network connect "kind" "${KIND_REGISTRY_NAME}"
+    fi
+
+    # Add the registry config to the nodes
+    #
+    # This is necessary because localhost resolves to loopback addresses that are network-namespace local.
+    # In other words: localhost in the container is not localhost on the host.
+    #
+    # We want a consistent name that works from both ends, so we tell containerd to alias localhost:${reg_port}
+    # to the registry container when pulling images
+    REGISTRY_DIR="/etc/containerd/certs.d/localhost:${KIND_REGISTRY_PORT}"
+    for node in $(kind get nodes); do
+        docker exec "${node}" mkdir -p "${REGISTRY_DIR}"
+        cat << EOF | docker exec -i "${node}" cp /dev/stdin "${REGISTRY_DIR}/hosts.toml"
+[host."http://${KIND_REGISTRY_NAME}:5000"]
+EOF
+    done
+
+    # Allow kind nodes to reach the registry
+    # docker network connect "kind" "${KIND_REGISTRY_NAME}"
+
+    # Document the local registry
+    cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: local-registry-hosting
+  namespace: kube-public
+data:
+  localRegistryHosting.v1: |
+    host: "localhost:${reg_port}"
+    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
+EOF
+}
+
+function build_and_push_images() {
+    HUB="${KIND_REGISTRY}" make docker.push
+}
+
+function install_dependencies() {
+    # 1. Install kind.
+    go install sigs.k8s.io/kind@v0.23.0
+}
+
+while (( "$#" )); do
+    case "$1" in
+    --skip-install-dep)
+      SKIP_INSTALL_DEPENDENCIES=true
+      shift
+    ;;
+    --skip-setup)
+      SKIP_SETUP=true
+      shift
+    ;;
+    --skip-build)
+      SKIP_BUILD=true
+      shift
+    ;;  
+    esac
+done
+
+if [[ -z "${SKIP_INSTALL_DEPENDENCIES:-}" ]]; then
+    install_dependencies
+fi
+
+if [[ -z "${SKIP_SETUP:-}" ]]; then
+    setup_kind_cluster
+fi
+
+if [[ -z "${SKIP_BUILD:-}" ]]; then
+    setup_kind_registry
+
+    build_and_push_images
+fi
+
+go test -v -tags=integ ./test/e2e/...
```

Comment:
```suggestion
go test -v -tags=integ $ROOT_DIR/test/e2e/...
```
ROOT_DIR=$(git rev-parse --show-toplevel)
Preventing test failures by not running in the kmesh root directory.